src/de.hpi.swa.trufflesqueak/src/de/hpi/swa/trufflesqueak/util/NFIUtils.java

```diff
@@ -0,0 +1,186 @@
+            this.executable = executable;
+            this.closure = createClosure(context, executable, executable.nfiSignature);
+        }
+    }
+
+    public interface ITruffleExecutable {
+        Object execute(Object... arguments);
+    }
+
+    @FunctionalInterface
+    public interface TruffleSupplier<R> extends ITruffleExecutable {
+        R run();
+
+        default Object execute(Object... arguments) {
+            assert arguments.length == 0;
+            return run();
+        }
+    }
+
+    @FunctionalInterface
+    public interface TruffleFunction<T, R> extends ITruffleExecutable {
+        R run(T argument);
+
+        default Object execute(Object... arguments) {
+            assert arguments.length == 1;
+            return run((T) arguments[0]);
+        }
+    }
+
+    @FunctionalInterface
+    public interface TruffleBiFunction<S, T, R> extends ITruffleExecutable {
+        R run(S argument1, T argument2);
+
+        default Object execute(Object... arguments) {
+            assert arguments.length == 2;
+            return run((S) arguments[0], (T) arguments[1]);
+        }
+    }
+
+    @FunctionalInterface
+    public interface TruffleTriFunction<S, T, U, R> extends ITruffleExecutable {
+        R run(S argument1, T argument2, U argument3);
+
+        default Object execute(Object... arguments) {
+            assert arguments.length == 3;
+            return run((S) arguments[0], (T) arguments[1], (U) arguments[2]);
+        }
+    }
+
+    @FunctionalInterface
+    public interface TruffleQuadFunction<S, T, U, V, R> extends ITruffleExecutable {
+        R run(S argument1, T argument2, U argument3, V argument4);
+
+        default Object execute(Object... arguments) {
+            assert arguments.length == 4;
+            return run((S) arguments[0], (T) arguments[1], (U) arguments[2], (V) arguments[3]);
+        }
+    }
+
+    @FunctionalInterface
+    public interface TruffleQuintFunction<S, T, U, V, W, R> extends ITruffleExecutable {
+        R run(S argument1, T argument2, U argument3, V argument4, W argument5);
+
+        default Object execute(Object... arguments) {
+            assert arguments.length == 5;
+            return run((S) arguments[0], (T) arguments[1], (U) arguments[2], (V) arguments[3], (W) arguments[4]);
+        }
+    }
+
+    public static Object executeNFI(SqueakImageContext context, String nfiCode) {
+        final Source source = Source.newBuilder("nfi", nfiCode, "native").build();
+        return context.env.parseInternal(source).call();
+    }
+
+    public static Object loadLibrary(SqueakImageContext context, String moduleName, String boundSymbols) {
+        final String libName = System.mapLibraryName(moduleName);
+        final TruffleFile libPath = context.getHomePath().resolve("lib" + File.separatorChar + libName);
```

Comment:
Are you sure that shared libraries are always prefixed with `lib`? I remember the file was called `SqueakSSL.so`, no? So maybe we should check for multiple naming conventions here?

Reply:
This was inspired by `SqueakFFIPrims.AbstractFFIPrimitiveNode#getPathOrFail`. I think prepending "lib" to the name of a shared librrary is the default for Cmake but disabled in the makefiles of opensmalltalk-vm.

Reply:
The FFI implementation in TruffleSqueak is rudimentary, so it's probably not a good idea to trust it. :wink: Instead, we should try and make sure that the plugins from OSVM work.